File: Samples/C#/VisualStyleAndVisibility/VisualStyleAndVisibilityPlugin.cs
/*  Sample: Model Object Quantities
 *
 *  This sample illustrates how to work with model object views, set visual style of the model view,
 *  hide model objects of specific type
 *
 *  Copyright Renga Software LLC, 2018. All rights reserved.
 */

using System.Text;
using System.Collections.Generic;

namespace VisualStyleAndVisibility
{
  public class VisualStyleAndVisibilityPlugin : Renga.IPlugin
  {
    private Renga.Application m_app;
    private Renga.ActionEventSource m_setNewVisualStyleToViewActionEventSource;
    private Renga.ActionEventSource m_hideAllWallsActionEventSource;

    public bool Initialize(string pluginFolder)
    {
      m_app = new Renga.Application();
      var ui = m_app.UI;
      var panelExtension = ui.CreateUIPanelExtension();

      {
        var action = ui.CreateAction();
        action.ToolTip = "Set new visual style to the view";
        m_setNewVisualStyleToViewActionEventSource = new Renga.ActionEventSource(action);
        m_setNewVisualStyleToViewActionEventSource.Triggered += (s, e) => SetNewVisualStyleToView();
        panelExtension.AddToolButton(action);
      }

      {
        var action = ui.CreateAction();
        action.ToolTip = "Hide all walls";
        m_hideAllWallsActionEventSource = new Renga.ActionEventSource(action);
        m_hideAllWallsActionEventSource.Triggered += (s, e) => HideAllWalls();
        panelExtension.AddToolButton(action);
      }

      ui.AddExtensionToPrimaryPanel(panelExtension);
      return true;
    }

    public void Stop()
    {
      m_setNewVisualStyleToViewActionEventSource.Dispose();
      m_hideAllWallsActionEventSource.Dispose();
    }

    private void SetNewVisualStyleToView()
    {
      var view = m_app.ActiveView;

      var modelView = view as Renga.IModelView;
      if (modelView == null)
        return;

      var visualStyle = modelView.VisualStyle;
      if (visualStyle != Renga.VisualStyle.VisualStyle_Color)
        visualStyle = Renga.VisualStyle.VisualStyle_Color;
      else
        visualStyle = Renga.VisualStyle.VisualStyle_Monochrome;

      modelView.VisualStyle = visualStyle;
    }
    private void HideAllWalls()
    {
      var view = m_app.ActiveView;

      var modelView = view as Renga.IModelView;
      if (modelView == null)
        return;

      var project = m_app.Project;
      var model = project.Model;
      var modelObjectCollection = model.GetObjects();


      var wallObjectIds = new List<int>();
      for (int i = 0; i < modelObjectCollection.Count; ++i)
      {
        var modelObject = modelObjectCollection.GetByIndex(i);
        if (modelObject.ObjectType == Renga.ObjectTypes.Wall)
          wallObjectIds.Add(modelObject.Id);
      }

      modelView.SetObjectsVisibility(wallObjectIds.ToArray(), false);
    }
  }
}

File: Samples/C++/VisualStyleAndVisibility/VisualStyleAndVisibilityPlugin.cpp
/*  Sample: Views
 *
 *  This sample illustrates how to work with model object views, set visual style of the model view,
 *  hide model objects of specific type
 *
 *  Copyright Renga Software LLC, 2018. All rights reserved.
 */

#include "VisualStyleAndVisibilityPlugin.h"

#include <sstream>
#include <vector>

#include <Renga/ObjectTypes.h>
#include <Renga/CreateApplication.hpp>

#include <atlsafe.h>


ViewsPlugin::ViewsPlugin()
{
  ::CoInitializeEx(nullptr, COINIT_APARTMENTTHREADED);
}

ViewsPlugin::~ViewsPlugin()
{
  ::CoUninitialize();
}

bool ViewsPlugin::initialize(const wchar_t* pluginDir)
{
  auto pApplication = Renga::CreateApplication();
  if (!pApplication)
    return false;

  if (auto pUI = pApplication->GetUI())
  {
    if (auto pUIPanelExtension = pUI->CreateUIPanelExtension())
    {
      if (auto pAction = pUI->CreateAction())
      {
        pAction->PutToolTip(L"Set new visual style to the view");
        m_buttonClickHandlers.emplace_back(new ViewVisualStyleButtonClickHandler(pAction, pApplication));
        pUIPanelExtension->AddToolButton(pAction);
      }

      if (auto pAction = pUI->CreateAction())
      {
        pAction->PutToolTip(L"Hide all walls");
        m_buttonClickHandlers.emplace_back(new HideAllWallsButtonClickHandler(pAction, pApplication));
        pUIPanelExtension->AddToolButton(pAction);
      }

      pUI->AddExtensionToPrimaryPanel(pUIPanelExtension);
    }
  }
  return true;
}

void ViewsPlugin::stop()
{
  m_buttonClickHandlers.clear();
}


ButtonClickHandler::ButtonClickHandler(Renga::IActionPtr pAction, Renga::IApplicationPtr pApplication) :
  Renga::ActionEventHandler(pAction),
  m_pApplication(pApplication)
{
}


ViewVisualStyleButtonClickHandler::ViewVisualStyleButtonClickHandler(Renga::IActionPtr pAction, Renga::IApplicationPtr pApplication) :
  ButtonClickHandler(pAction, pApplication)
{
}

void ViewVisualStyleButtonClickHandler::OnTriggered()
{
  auto pView = m_pApplication->GetActiveView();

  Renga::IModelViewPtr pModelView;
  pView->QueryInterface(&pModelView);

  if (!pModelView)
    return;

  auto visualStyle = pModelView->GetVisualStyle();
  if (visualStyle != Renga::VisualStyle::VisualStyle_Color)
    visualStyle = Renga::VisualStyle::VisualStyle_Color;
  else
    visualStyle = Renga::VisualStyle::VisualStyle_Monochrome;

  pModelView->PutVisualStyle(visualStyle);
}


HideAllWallsButtonClickHandler::HideAllWallsButtonClickHandler(Renga::IActionPtr pAction, Renga::IApplicationPtr pApplication) :
  ButtonClickHandler(pAction, pApplication)
{
}

void HideAllWallsButtonClickHandler::OnTriggered()
{
  auto pView = m_pApplication->GetActiveView();

  Renga::IModelViewPtr pModelView;
  pView->QueryInterface(&pModelView);

  if (!pModelView)
    return;

  auto pProject = m_pApplication->GetProject();
  auto pModel = pProject->GetModel();
  auto pModelObjectCollection = pModel->GetObjects();

  std::vector<int> wallObjectIds;
  for (int i = 0; i < pModelObjectCollection->GetCount(); ++i)
  {
    auto pModelObject = pModelObjectCollection->GetByIndex(i);
    if (pModelObject->GetObjectType() == Renga::ObjectTypes::Wall)
      wallObjectIds.push_back(pModelObject->GetId());
  }

  CComSafeArray<int> wallObjectIdsSafeArray(static_cast<ULONG>(wallObjectIds.size()));
  for (int i = 0; i < wallObjectIds.size(); i++)
    wallObjectIdsSafeArray.SetAt(i, wallObjectIds[i]);

  pModelView->SetObjectsVisibility(wallObjectIdsSafeArray, FALSE);
}
